### PR TITLE
feat: compute state root for #7784

### PR DESCRIPTION
**Pull request by onbjerg**
- Computes the state root for the imported state dump
- Compares it to the state root in the state dump (if any)
- Also creates one provider only in order to commit once to preserve integrity

### CODE REVIEW

File: crates/node-core/src/init.rs [hunk 1]
```diff
@@ -379,10 +379,40 @@ pub fn init_from_state_dump<DB: Database>(
...
+/// Computes the state root (from scratch) based on the accounts and storages present in the
+/// database.
+fn compute_state_root<DB: Database>(provider: &DatabaseProviderRW<DB>) -> eyre::Result<StateRoot> {
+    trace!(target: "reth::cli", "Computing state root");
+
+    let tx = provider.tx_ref();
+    let (root, updates) = StateRootComputer::from_tx(tx)
+        .with_intermediate_state(None)
+        .with_no_threshold()
+        .root_with_updates()?;
```
Comment (rkrasiuk) [inline crates/node-core/src/init.rs:409]: nit: `root_with_updates` already sets `with_no_threshold` and `with_intermediate_state(None)` is a noop
```suggestion
        .root_with_updates()?;
```

**Review by rkrasiuk — COMMENTED** [review]
i have oom concerns as all intermediate nodes will be aggregated in memory until getting flushed

**Comment by onbjerg** [comment]
@rkrasiuk unfortunately we are replicating pipeline logic here (again) and doing it over multiple runs is really iffy with how it's set up right now. i think we decided internally that oom is not a concern for us right now because this is op specific, but yes, i agree

is it possible to flush as we go while still having no threshold?

**Comment by onbjerg** [comment]
now flushes trie updates every once in a while h/t @rkrasiuk

File: crates/node-core/src/init.rs [hunk 2]
```diff
@@ -379,10 +379,53 @@ pub fn init_from_state_dump<DB: Database>(
...
+
+    let tx = provider.tx_ref();
+    let mut intermediate_state: Option<IntermediateStateRootState> = None;
+    let root = loop {
+        match StateRootComputer::from_tx(tx)
+            .with_intermediate_state(intermediate_state)
+            .root_with_progress()?
+        {
+            StateRootProgress::Progress(state, _, updates) => {
+                trace!(target: "reth::cli", "Flushing trie updates");
```
Comment (rkrasiuk) [inline crates/node-core/src/init.rs:413]:
```suggestion
                trace!(target: "reth::cli", last_account_key = %state.last_account_key, updates_len = updates.len(), "Flushing trie updates");
```

File: crates/node-core/src/init.rs [hunk 3]
```diff
@@ -379,10 +379,53 @@ pub fn init_from_state_dump<DB: Database>(
...
+                intermediate_state = Some(*state);
+                updates.flush(tx)?;
+            }
+            StateRootProgress::Complete(root, _, updates) => {
+                trace!(target: "reth::cli", "State root has been computed");
```
Comment (rkrasiuk) [inline crates/node-core/src/init.rs:418]:
```suggestion
                trace!(target: "reth::cli", updates_len = updates.len(), "State root has been computed");
```

**Review by rkrasiuk — APPROVED** [review]
observability nits, otherwise lgtm!

File: crates/node-core/src/init.rs (hunk 2 above)
Comment (rkrasiuk) [reply crates/node-core/src/init.rs:413]: @onbjerg last_account_key would give you visibility into the progress of the computation

**Review by joshieDo — APPROVED** [review]

**Review by emhane — APPROVED** [review]

**Pull request merged**